### ISSUE TITLE
macro_rules: incorrect warning "unreachable_patterns"

### ISSUE THREAD

**Issue by Vernoxvernax** [rust-lang/rust#124937]
I tried this code:

```rust
macro_rules! SomeMacro {
    ($(($var:ident, $str1:literal, $str2:literal, $str3:literal),)+) => {
        pub enum SomeMacro {
            $($var,)*
        }

        impl SomeMacro {
            fn try_from(s: &str) {
                match s {
                    $($str1 | $str2 | $str3 => println!("hihi"),)+
                    _ => println!("haha"),
                }
            }
        }
    }
}

SomeMacro! {
    (A, "aa", "aa", "aaa"),
    (B, "b", "bb", "bbb"),
    (C, "c", "cc", "ccc"),
    (D, "d", "dd", "ddd"),
}
```

When compiling I get a warning that the `$str2` is an unreachable pattern.
That is correct for the first line in `SomeMacro!`, but anything after could technically require the second arm as well.
It's my first time playing around with macros, so I apologize in advance, if this is expected behaviour.

`rustc --version --verbose`:
```
rustc 1.78.0 (9b00956e5 2024-04-29)
binary: rustc
commit-hash: 9b00956e56009bab2aa15d7bff10916599e3d6d6
commit-date: 2024-04-29
host: x86_64-unknown-linux-gnu
release: 1.78.0
LLVM version: 18.1.2
```
I also tested it with the nightly channel.
and sorry if the tag is wrong.

**Comment by Kaida-Amethyst** [comment]
I find if you use :

```rust
SomeMacro! {
    (A, "a", "aa", "aaa"),   // <-- change this line.
    (B, "b", "bb", "bbb"),
    (C, "c", "cc", "ccc"),
    (D, "d", "dd", "ddd"),
}
```

The warning would disappear. ^^

**Comment by Vernoxvernax** [comment]
> I find if you use :
> 
> ```rust
> SomeMacro! {
>     (A, "a", "aa", "aaa"),   // <-- change this line.
>     (B, "b", "bb", "bbb"),
>     (C, "c", "cc", "ccc"),
>     (D, "d", "dd", "ddd"),
> }
> ```
> 
> The warning would disappear. ^^

Thank you for your reply.
I was already aware of that fact. Are there any rules about the content of macros, that would make my code sample invalid? I thought what I was doing, inserting the same content into the second literal, could be expected.

Context: A macro for language codes could sometimes have the same content in two literals:

```rust
...
(ENG, "eng", "eng"),
(GER, "deu", "ger"),
...
```
The first literal being the language code in it's native language (Deutsch).

**Comment by Kaida-Amethyst** [comment]
May be you could try to use "guard"?

```rust
macro_rules! SomeMacro {
    ($(($var:ident, $str1:literal, $str2:literal, $str3:literal),)+) => {
        pub enum SomeMacro {
            $($var,)*
        }

        impl SomeMacro {
            fn try_from(s: &str) {
                match s {
                    $(st if st == $str1 || st == $str2 || st == $str3 => println!("hihi"),)+
                    _ => println!("haha"),
                }
            }
        }
    }
}
```

^^ docs here: https://doc.rust-lang.org/rust-by-example/flow_control/match/guard.html

**Comment by Vernoxvernax** [comment]
> May be you could try to use "guard"?
> 
> ```rust
> macro_rules! SomeMacro {
>     ($(($var:ident, $str1:literal, $str2:literal, $str3:literal),)+) => {
>         pub enum SomeMacro {
>             $($var,)*
>         }
> 
>         impl SomeMacro {
>             fn try_from(s: &str) {
>                 match s {
>                     $(st if st == $str1 || st == $str2 || st == $str3 => println!("hihi"),)+  // <----change this
>                     _ => println!("haha"),
>                 }
>             }
>         }
>     }
> }
> ```
> 
> ^^ docs here: https://doc.rust-lang.org/rust-by-example/flow_control/match/guard.html

Yup, that does it. Thank you very much.
I don't know, why I didn't think of trying a different syntax ...

I'll keep this issue open until someone confirms that it's expected behaviour.

**Comment by jieyouxu** [comment]
> I was already aware of that fact. Are there any rules about the content of macros, that would make my code sample invalid? I thought what I was doing, inserting the same content into the second literal, could be expected.

`(A, "aa", "aa", "aaa")` would expand to

```rs
match s {
    "aa" | "aa" | "aaa" => println!("hihi"),
    _ => println!("haha"),
}
```

which is exactly what `unreachable_patterns` is designed to catch AFAICT.

**Comment by jieyouxu** [comment]
> I see. I just thought that the warning should only appear when it actually applies to the entire content of the macro.

I don't quite understand what you mean by "actually applies to the entire content of the macro"? `unreachable_patterns` warns for *any* instance of, well, unreachable patterns.

**Comment by Vernoxvernax** [comment]
> I don't quite understand what you mean by "actually applies to the entire content of the macro"? `unreachable_patterns` warns for _any_ instance of, well, unreachable patterns.

I just wanted to say that I was confused by the warning.
Sure, it makes sense that the warning pops up, even if it matches for only one line.
This is kinda my introduction into macros and I keep thinking they work the same as vectors which is clearly not the case.
Sorry for the trouble

**Comment by jieyouxu** [comment]
> I just wanted to say that I was confused by the warning. Sure, it makes sense that the warning pops up, even if it matches for only one line. This is kinda my introduction into macros and I keep thinking they work the same as vectors which is clearly not the case. Sorry for the trouble

Right, the warning can be a bit confusing because it can't refer to the expanded code. This expanded code doesn't "exist" in your source file and has to refer to what *does* exist, i.e. the macro source. No need to be sorry especially when it's your introduction to macros.
